lessons/00-intro/best-practices/best_practices_guide.md

```diff
@@ -0,0 +1,260 @@
+- Critical packages are installed on the development server, but are missing on the production server
+- Conflicts in library versions among different projects
+
+In order to avoid these embarrassing situations, the project configuration needs to be explicit so that results are reproducible (on a local or remote server, by any developer, ...).
+
+> That's why any project must come with its virtual environment that is responsible for managing the Python version, as well as all the dependencies.
+
+A virtual environment is always summarized in a requirements.txt (pip-fashion) or a environment.yml (conda-fashion) file, located at the root of the project. It takes the following form:
+
+- For `requirements.txt`
+
+```
+pandas==1.1.0
+```
+
+- For `environment.yml`
+
+```
+name: envname
+channels:
+  - conda-forge
+dependencies:
+  - pandas=1.1.0
+```
+
+### Setting up a virtual environment
+
+We will present 2 options: using `virtualenv` or `conda`
+
+#### Virtualenv
+
+```
+# Install virtualenv package
+$ pip install virtualenv
+
+# Create the virtual environment
+$ virtualenv venv --python=python3.x
+
+# Activate it
+$ source venv/Scripts/activate  # Windows
+$ source venv/bin/activate  # Unix
+```
+
+#### Conda
+
+If you have a Miniconda installed:
+
+```
+# Create env
+$ cd <this repository>
+$ conda env create --file environment.yml
+
+# Activate your environment:
+$ conda activate envname
+
+# Check python:
+(envname) $ which python # Unix
+(envname) $ where python # Windows
+
+# In order to stop using this virtual environment:
+(envname) $ deactivate
+```
+
+Once activated, the command line starts with (`envname`) to let you know which environment you're in. The environment must always be activated while working in the project.
+
+### Handle dependencies
+
+In your base environment, there should not be any Python packages. This isolates the dependencies of any project, avoiding future compatibility problems. Therefore, you must always install dependencies from within your virtual environment.
+
+#### Virtualenv
+
+```
+$ source venv/bin/activate  # Activate env
+(venv) $ pip install -r requirements.txt
+```
+
+If you need to amend the dependencies of the project, then simply run:
+
+```
+(venv) $ pip install mypackage
+# You can have an exhaustive view of your environment:
+(venv) $ pip list | grep pandas
+pandas (1.1.0)
+```
+
+You should not forget to add this new dependence to `requirements.txt`. Our recommended way is to have a `requirements.in`:
+
+- Just add a line in the file, specifying the version (example: `pandas==1.0.1`)
+- Overwrite `requirements.txt` file by re-generating it:
+
+```
+(venv) $ pip-compile requirements.in
+```
+
+This will automatically parse all your nested dependencies into `requirements.txt` thus ensuring full reproducibility. The only caveats are that it requires an additional package, `pip-tools` and that it takes some time to compile long and complex requirements.
```

Comment:
I think here you lack a sentence about what dependencies go to requirements.in (the ones called in your python libs)